### PR TITLE
[InstCombine] Handle isNanOrInf idioms

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch folds:
```
(icmp eq (and (bitcast X to int), ExponentMask), ExponentMask) --> llvm.is.fpclass(X, fcInf|fcNan)
(icmp ne (and (bitcast X to int), ExponentMask), ExponentMask) --> llvm.is.fpclass(X, ~(fcInf|fcNan))
```
Alive2: https://alive2.llvm.org/ce/z/_hXAAF

Related patch: https://github.com/llvm/llvm-project/pull/76338

### CODE REVIEW

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+; CHECK-NEXT:    [[AND:%.*]] = and i32 [[I32]], 2139095040
+; CHECK-NEXT:    [[CMP:%.*]] = icmp eq i32 [[AND]], 2139095040
+; CHECK-NEXT:    ret i1 [[CMP]]
+;
+  %i32 = bitcast float %a to i32
+  %and = and i32 %i32, 2139095040
+  %cmp = icmp eq i32 %and, 2139095040
+  ret i1 %cmp
+}
+
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:176]: can you add a strictfp test? Also some with a vector result

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -1827,6 +1827,35 @@ Instruction *InstCombinerImpl::foldICmpAndConstConst(ICmpInst &Cmp,
...
+          Attribute::NoImplicitFloat) &&
+      Cmp.isEquality() && match(X, m_OneUse(m_BitCast(m_Value(V))))) {
+    Type *SrcType = V->getType();
+    Type *DstType = X->getType();
+    // Make sure the bitcast doesn't change between scalar and vector and
+    // doesn't change the number of vector elements.
+    if (SrcType->isVectorTy() == DstType->isVectorTy() &&
+        SrcType->getScalarSizeInBits() == DstType->getScalarSizeInBits()) {
+      Type *FPType = SrcType->getScalarType();
+      if (FPType->isIEEELikeFPTy() && C1 == *C2) {
```
Comment (arsenm) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:1845]: Can make this one big if ?Just hoist the FPType def?

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll (hunk 1 above)
Comment (dtcxzyw) [reply llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:202]: Done.

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 3]
```diff
@@ -1827,6 +1827,35 @@ Instruction *InstCombinerImpl::foldICmpAndConstConst(ICmpInst &Cmp,
...
+        APInt ExponentMask = APInt::getBitsSet(
+            FPType->getScalarSizeInBits(), FPType->getFPMantissaWidth() - 1,
+            FPType->getScalarSizeInBits() - 1);
```
[the remark this answers is not in the document]
Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp]: Use `APFloat::getInf(FPType->getFltSemantics()).bitcastToAPInt()` instead.

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll [hunk 4]
```diff
@@ -75,6 +75,32 @@ define i1 @f32_fcnan(float %a) {
   ret i1 %res
 }
 
+define i1 @f32_fcnan_fcinf_strictfp(float %a) strictfp {
+; CHECK-LABEL: define i1 @f32_fcnan_fcinf_strictfp(
+; CHECK-SAME: float [[A:%.*]]) #[[ATTR0:[0-9]+]] {
+; CHECK-NEXT:    [[TMP1:%.*]] = call float @llvm.fabs.f32(float [[A]])
+; CHECK-NEXT:    [[CMP:%.*]] = fcmp ueq float [[TMP1]], 0x7FF0000000000000
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:82]: this shows a bug, we're not emitting a strict compare. doesn't need to be fixed in this patch though, existing issue

**Review by arsenm — APPROVED** [review]

**Comment by andykaylor** [comment]
@dtcxzyw FYI, this patch and your similar recent changes introduce runtime library calls with fp128 (https://godbolt.org/z/PWcjK86vz) so it's a bit of a de-optimization in that case. I think @FreddyLeaf will look into this when he returns from holiday. We should be able to fix this in the selection DAG.

@pranavk has been working on fp128-related issues, so may also want to be aware of this.

**Comment by dtcxzyw** [comment]
> @dtcxzyw FYI, this patch and your similar recent changes introduce runtime library calls with fp128 (https://godbolt.org/z/PWcjK86vz) so it's a bit of a de-optimization in that case. I think @FreddyLeaf will look into this when he returns from holiday. We should be able to fix this in the selection DAG.
> 
> @pranavk has been working on fp128-related issues, so may also want to be aware of this.

I will reverse the transform in SDAG to fix the problem.
See also https://github.com/llvm/llvm-project/pull/76338#issuecomment-1929678129.

**Comment by andykaylor** [comment]
> I will reverse the transform in SDAG to fix the problem. See also [#76338 (comment)](https://github.com/llvm/llvm-project/pull/76338#issuecomment-1929678129).

Thanks! This broke the build of some downstream libraries I work with because they didn't want to link with the runtime library. We disabled the transformations for fp128 locally to work around the problem, but I think this is a good transformation and I'd like to re-enable it when we can.

You may already be considering this, but x86_fp80 also has issues, in that after your changes it's using x87 instructions where it used to use a faster integer implementation: https://godbolt.org/z/bPfrnW368

File: llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll (hunk 4 above)
Comment (ostannard) [reply llvm/test/Transforms/InstCombine/fpclass-check-idioms.ll:82]: I think this is an old bug but this patch has caused us to start hitting it in our floating-point library, I've posted a fix at #81498.

**Pull request merged**